lib/Extractor/Solver.cpp

```diff
@@ -472,6 +472,94 @@ class BaseSolver : public Solver {
     return !IsSat;
   }
 
+  std::error_code testRange(const BlockPCs &BPCs,
+                            const std::vector<InstMapping> &PCs,
+                            Inst *LHS, llvm::APInt &C,
+                            llvm::APInt &X,
+                            bool &IsFound,
+                            InstContext &IC) {
+
+
+    unsigned W = LHS->Width;
+
+    Inst *SynthesisX = IC.createSynthesisConstant(W, 1);
+    Inst *CVal = IC.getConst(C);
+    Inst *LowerVal = SynthesisX;
+    Inst *UpperValOverflow = IC.getInst(Inst::UAddWithOverflow, W + 1,
+                                        {IC.getInst(Inst::Add, W, {LowerVal, CVal}),
+                                         IC.getInst(Inst::UAddO, 1, {LowerVal, CVal})});
+
+    Inst *IsOverflow = IC.getInst(Inst::ExtractValue, 1, {UpperValOverflow, IC.getUntypedConst(llvm::APInt(W, 1))});
+    Inst *UpperVal = IC.getInst(Inst::ExtractValue, W, {UpperValOverflow, IC.getUntypedConst(llvm::APInt(W, 0))});
+
+    Inst *GuessLowerPartNonWrapped = IC.getInst(Inst::Ule, 1, {LowerVal, LHS});
+    Inst *GuessUpperPartNonWrapped = IC.getInst(Inst::Ult, 1, {LHS, UpperVal});
+
+    // non-wrapped, x <= LHS < x+c
+    Inst *GuessAnd = IC.getInst(Inst::And, 1, { GuessLowerPartNonWrapped, GuessUpperPartNonWrapped });
+    // wrapped, LHS < x+c \/ LHS >= x
+    Inst *GuessOr = IC.getInst(Inst::Or, 1, { GuessLowerPartNonWrapped, GuessUpperPartNonWrapped });
+
+    // if x+c overflows, treat it as wrapped.
+    Inst *Guess = IC.getInst(Inst::Select, 1, {IsOverflow, GuessOr, GuessAnd});
+
+
+    std::set<Inst *> ConstSet{SynthesisX};
+    std::map <Inst *, llvm::APInt> ResultMap;
+    ConstantSynthesis CS;
+    std::error_code EC;
+    EC = CS.synthesize(SMTSolver.get(), BPCs, PCs, InstMapping(Guess, IC.getConst(APInt(1, true))),
+                       ConstSet, ResultMap, IC, /*MaxTries=*/30, Timeout);
+    if (ResultMap.empty()) {
+      IsFound = false;
+    } else {
+      IsFound = true;
+      X = ResultMap[SynthesisX];
+    }
+
+    return EC;
+  }
+
+  std::error_code range(const BlockPCs &BPCs,
+                        const std::vector<InstMapping> &PCs,
+                        Inst *LHS, llvm::ConstantRange &Range,
+                        InstContext &IC) override {
+    std::error_code EC;
+    unsigned W = LHS->Width;
+
+    APInt L = APInt(W, 1), R = APInt::getAllOnesValue(W);
+    APInt BinSearchResultX, BinSearchResultC;
+    bool BinSearchHasResult = false;
+
+    while (L.ule(R)) {
+      APInt M = L + ((R - L)).lshr(1);
+      APInt BinSearchX;
+      bool Found = false;
+      EC = testRange(BPCs, PCs, LHS, M, BinSearchX, Found, IC);
+      if (EC) {
+        llvm::report_fatal_error("stopping due to error");
```

Comment:
does this ever happen in practice?

Reply:
Sorry, it wont happen, after calling testRange, either we found the range or we don't. I'll remove this line.